### PR TITLE
feat: throttle price tick publishing

### PR DESCRIPTION
**Pull request by enigbe**
Throttles the publishing of price ticks

### CODE REVIEW

File: shared/src/pubsub/publisher.rs [hunk 1]
```diff
@@ -18,12 +30,43 @@ impl Publisher {
...
-        Ok(Self { client })
+        Ok(Self {
+            client,
+            rate_limiter: Arc::new(RateLimiter::keyed(Quota::per_minute(
+                NonZeroU32::new(RATE_LIMIT).unwrap(),
+            ))),
+        })
+    }
+
+    pub fn rate_limit_publisher(
```
Comment (bodymindarts) [inline shared/src/pubsub/publisher.rs:41]: why pub?

File: shared/src/pubsub/publisher.rs [hunk 2]
```diff
@@ -18,12 +30,43 @@ impl Publisher {
...
+    ) -> Result<bool, PublisherError> {
+        match self.rate_limit_publisher(<P as MessagePayload>::message_type()) {
+            Ok(publisher) => {
+                let publish_result = publisher.publish(payload).await;
+                match publish_result {
+                    Ok(_res) => Ok(true),
+                    Err(err) => Err(err),
+                }
+            }
+            Err(_) => {
```
Comment (bodymindarts) [inline shared/src/pubsub/publisher.rs:62]:
```suggestion
        if let Ok(_) = self.rate_limiter.check_key(&key) {
        self.publish.await?;
        Ok(true)
        } else {
          Ok(false)
        }
```

File: shared/src/pubsub/publisher.rs [hunk 3]
```diff
@@ -18,12 +30,43 @@ impl Publisher {
...
+        })
+    }
+
+    pub fn rate_limit_publisher(
+        &self,
+        key: &'static str,
+    ) -> Result<&Publisher, NotUntil<QuantaInstant>> {
+        self.rate_limiter.check_key(&key)?;
+        Ok(self)
+    }
```
Comment (bodymindarts) [inline shared/src/pubsub/publisher.rs:47]:
```suggestion

```

Comment (enigbe) [reply shared/src/pubsub/publisher.rs]: Your earlier suggestion was right. This should not be public.

File: shared/src/pubsub/publisher.rs [hunk 4]
```diff
@@ -1,29 +1,68 @@
...
         client
             .wait_for_connect()
             .await
             .map_err(PublisherError::InitialConnection)?;
-        Ok(Self { client })
+
+        Ok(Self {
+            client,
+            rate_limiter: Arc::new(RateLimiter::keyed(
+                Quota::with_period(Duration::from_secs(config.rate_limit_interval.unwrap()))
```
Comment (bodymindarts) [inline shared/src/pubsub/publisher.rs:37]: how is `unwrap` guaranteed to work?

File: shared/src/pubsub/publisher.rs [hunk 5]
```diff
@@ -1,29 +1,68 @@
...
+                    .unwrap()
+                    .allow_burst(NonZeroU32::new(MAX_BURST).unwrap()),
+            )),
+        })
+    }
+
+    fn rate_limit_publisher(
```
Comment (bodymindarts) [inline shared/src/pubsub/publisher.rs:44]: What is this function for? Why not inline `self.rate_limiter.check_key(&key)`?

File: shared/src/pubsub/publisher.rs [hunk 6]
```diff
@@ -1,29 +1,68 @@
...
             .wait_for_connect()
             .await
             .map_err(PublisherError::InitialConnection)?;
-        Ok(Self { client })
+
+        Ok(Self {
+            client,
+            rate_limiter: Arc::new(RateLimiter::keyed(
+                Quota::with_period(Duration::from_secs(config.rate_limit_interval.unwrap()))
+                    .unwrap()
```
Comment (bodymindarts) [inline shared/src/pubsub/publisher.rs:38]: We should never use unwrap in production code - always `expect`

**Pull request merged**